hazelcast/src/main/java/com/hazelcast/internal/networking/nio/NioPipeline.java

```diff
@@ -136,36 +137,127 @@ final void unregisterOp(int operation) throws IOException {
+     * IOException should not be dealt with in the handle method but should
+     * be propagated. The reason behind this is that the handle logic already
+     * is complicated enough and by pulling it out the flow will be easier to
+     * understand.
      *
      * @throws Exception
      */
-    public abstract void process() throws Exception;
+    abstract void process() throws Exception;
+
+    /**
+     * Forces this NioPipeline to wakeup by scheduling this on the owner using
+     * {@link NioThread#addTaskAndWakeup(Runnable)}
+     *
+     * This method can be called by any thread. It is a pretty expensive method
+     * because it will cause the {@link Selector#wakeup()} method to be called.
+     *
+     * This call can safely be made during the migration of the pipeline.
+     */
+    final void wakeup() {
+        addTaskAndWakeup(this);
+    }
+
+    /**
+     * Adds a task to be executed on the {@link NioThread owner}.
+     *
+     * This task is scheduled on the task queue of the owning {@link NioThread}.
+     *
+     * If the pipeline is currently migrating, this method will make sure the
+     * task ends up at the new owner.
+     *
+     * It is extremely important that this task takes very little time because
+     * otherwise it could cause a lot of problems in the IOSystem.
+     *
+     * This method can be called by any thread. It is a pretty expensive method
+     * because it will cause the {@link Selector#wakeup()} method to be called.
+     *
+     * @param task the task to add.
+     */
+    final void addTaskAndWakeup(Runnable task) {
+        // in this loop we are going to either send the task to the owner
+        // or store the delayed task to be picked up as soon as the migration
+        // completes
+        for (; ; ) {
+            NioThread localOwner = owner;
+            if (localOwner != null) {
+                // there is an owner, lets send the task.
+                localOwner.addTaskAndWakeup(task);
+                return;
+            }
+
+            // there is no owner, so we put the task on the delayedTaskStack
+            TaskNode old = delayedTaskStack.get();
+            TaskNode update = new TaskNode(task, old);
+            if (delayedTaskStack.compareAndSet(old, update)) {
+                break;
+            }
+        }
+
+        NioThread localOwner = owner;
+        if (localOwner != null) {
+            // an owner was set, but we have no guarantee that he has seen our task.
+            // So lets try to reschedule the delayed tasks to make sure they get executed.
+            restoreTasks(localOwner, delayedTaskStack.getAndSet(null), true);
+        }
+    }
+
+    private void restoreTasks(NioThread owner, TaskNode node, boolean wakeup) {
+        if (node == null) {
+            return;
+        }
+
+        // we restore in the opposite order so that we get fifo.
+        restoreTasks(owner, node.next, false);
```

Comment:
`wakeup == false`, so only the last `task` will wakeup the selector/thread, right?

Reply:
Correct.